### PR TITLE
Make sure group searches are cached and entries invalidate accordingly

### LINKED ISSUE #23056: Flaky test: org.keycloak.testsuite.admin.concurrency.ConcurrencyTest#testAllConcurrently
Successor to #16155, might be a new issue?

## org.keycloak.testsuite.admin.concurrency.ConcurrencyTest#testAllConcurrently
[Keycloak CI - Base IT (1)](https://github.com/keycloak/keycloak/actions/runs/6087431274/job/16516434085) / [Pull Request #22824](https://github.com/keycloak/keycloak/pull/22824)

```
java.lang.RuntimeException: There were failures in threads. Failures count: 1
	at org.keycloak.testsuite.admin.concurrency.AbstractConcurrencyTest.run(AbstractConcurrencyTest.java:122)
	at org.keycloak.testsuite.admin.concurrency.AbstractConcurrencyTest.run(AbstractConcurrencyTest.java:63)
	at org.keycloak.testsuite.admin.concurrency.AbstractConcurrencyTest.run(AbstractConcurrencyTest.java:59)
	at org.keycloak.testsuite.admin.concurrency.ConcurrencyTest.concurrentTest(ConcurrencyTest.java:60)
...
```

Checked a couple runs and it's createGroups that seems to be the issue, which is linked to https://github.com/keycloak/keycloak/issues/24337 and https://github.com/keycloak/keycloak/issues/26983

### LINKED ISSUE #26983: Group not found in list after creation
### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

core

### Describe the bug

ConcurrencyTest#createGroup started failing on the 27th October 2023, which was the day after https://github.com/keycloak/keycloak/commit/69497382d8cc6c622e815184d9bc27cb0dc6a400 was merged.

The error is that a newly created group is not found in the group list:
```
2024-02-13T07:58:05.2629139Z ConcurrencyTest ++ java.lang.AssertionError: Group g-79 [affcccbf-f981-43a0-8f85-48dd664a9c87]  not found in group list
2024-02-13T07:58:05.2630974Z ConcurrencyTest ++ 	at org.junit.Assert.fail(Assert.java:89)
2024-02-13T07:58:05.2631978Z ConcurrencyTest ++ 	at org.junit.Assert.assertTrue(Assert.java:42)
2024-02-13T07:58:05.2633037Z ConcurrencyTest ++ 	at org.junit.Assert.assertFalse(Assert.java:65)
2024-02-13T07:58:05.2634509Z ConcurrencyTest ++ 	at org.keycloak.testsuite.admin.concurrency.ConcurrencyTest$CreateGroup.run(ConcurrencyTest.java:248)
2024-02-13T07:58:05.2637014Z ConcurrencyTest ++ 	at org.keycloak.testsuite.admin.concurrency.AbstractConcurrencyTest.lambda$run$0(AbstractConcurrencyTest.java:94)
2024-02-13T07:58:05.2638885Z ConcurrencyTest ++ 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
2024-02-13T07:58:05.2640522Z ConcurrencyTest ++ 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
2024-02-13T07:58:05.2642345Z ConcurrencyTest ++ 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
2024-02-13T07:58:05.2643793Z ConcurrencyTest ++ 	at java.base/java.lang.Thread.run(Thread.java:840)
```

The mentioned PR also introduced a hack to retry listing the groups after the group was created:
https://github.com/keycloak/keycloak/blob/ee126f73e532baaabd8c82b94df8c70031ff650b/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/admin/concurrency/ConcurrencyTest.java#L238-L248

However, this should not be needed the group should be available immediately. As a side-note the retry mechanism above has no wait in-between the retries, so will just pull the endpoint many times in a row.

### Version

main

### Expected behavior

Groups are available in the list after they have been created

### Actual behavior

Groups are not available in the list even after retrying several times

### How to Reproduce?

Run ConcurrencyTest.createGroup, or view results from https://github.com/keycloak/keycloak/issues/24337

### Anything else?

_No response_

### LINKED ISSUE #27095: Cache Keys for Group pagination and other entries cannot be invalidated and updated
### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

infinispan

### Describe the bug

While exploring a solution for #26983 that preserves the ability to cache paginated entries a bug revealed itself to me. It is possible currently to add a a cache entry for some search+pagination combination but then that entry will never be invalidated when any group changes happen. 

Currently the underlying group IDs will have their cache entries invalidated and then updated accordingly but any new group that gets added to the cache will not necessarily reflect in the results properly due to the cache entry never being invalidated/updated with the new/removed group IDs. 

This is technically a regression BUT we should preserve the paginated/search based cache entries

### Version

Main

### Regression

- [X] The issue is a regressions

### Expected behavior

Cache entry updates as expected

### Actual behavior

Cache entry will have stale group IDs

### How to Reproduce?

Perform the same paginated search query on groups multiple times while changing the contents of that page

### Anything else?

I have a suggested fix for this issue that I will link

### PR DESCRIPTION
**Pull request by pedroigor**
Closes #27095
Closes #23056
Closes #26983

### CODE REVIEW

File: server-spi/src/main/java/org/keycloak/models/GroupProvider.java [hunk 1]
```diff
@@ -205,6 +204,8 @@ default GroupModel createGroup(RealmModel realm, String name, GroupModel toParen
      */
     GroupModel createGroup(RealmModel realm, String id, String name, GroupModel toParent);
 
+    Set<String> getTopLevelGroupsIdsStream(RealmModel realm, String search, Boolean exact, Integer firstResult, Integer maxResults);
```
Comment (martin-kanis) [inline server-spi/src/main/java/org/keycloak/models/GroupProvider.java:207]: Please keep it consistent with other methods and return Stream<String>. Also please add javadoc.

File: model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/RealmCacheSession.java [hunk 2]
```diff
@@ -1016,40 +1016,62 @@ public Stream<GroupModel> getGroupsByRoleStream(RealmModel realm, RoleModel role
...
+                logger.tracev("adding realm getTopLevelGroups search cache miss: realm {0} key {1}", realm.getName(), searchKey);
+                cache.addRevisioned(query, cache.getCurrentCounter());
             }
-            list.add(group);
         }
 
-        return list.stream().sorted(GroupModel.COMPARE_BY_NAME);
+        List<GroupModel> groups = cached.stream()
+                .map((id) -> session.groups().getGroupById(realm, id))
+                .takeWhile(Objects::nonNull)
```
Comment (martin-kanis) [inline model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/RealmCacheSession.java:1054]: Could this perform better (by not consuming the stream)?

```
AtomicBoolean invalidate = new AtomicBoolean(false);
        Stream<GroupModel> groups = cached.stream()
                .map((id) -> session.groups().getGroupById(realm, id))
                .takeWhile(group -> {
                    if (Objects.isNull(group)) {
                        invalidate.set(true);
                        return false;
                    }
                    return true;
                })
                .sorted(GroupModel.COMPARE_BY_NAME);

        if (!invalidate.get()) {
            return groups;
        }
```

**Review by martin-kanis — CHANGES_REQUESTED** [review]
@pedroigor Thanks for this proposal. It looks good to me. It works and also fixes the test. Please see the following comments:

File: model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/RealmCacheSession.java [hunk 3]
```diff
@@ -1016,40 +1016,62 @@ public Stream<GroupModel> getGroupsByRoleStream(RealmModel realm, RoleModel role
...
+            return groups.stream();
+        }
+
+        invalidations.add(cacheKey);
+
+        return getGroupDelegate().getTopLevelGroupsStream(realm);
+    }
+
+    @Override
+    public Set<String> getTopLevelGroupsIdsStream(RealmModel realm, String search, Boolean exact, Integer firstResult, Integer maxResults) {
```
Comment (pedroigor) [inline model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/RealmCacheSession.java:1068]: Remove `Stream` from the method name.

File: server-spi/src/main/java/org/keycloak/models/GroupProvider.java (hunk 1 above)
Comment (pedroigor) [reply server-spi/src/main/java/org/keycloak/models/GroupProvider.java]: I've removed this method in favor of not breaking backward compatibility. We can look at it as a follow-up and see how to improve all this back-and-forth when dealign with streams later.

**Review by martin-kanis — APPROVED** [review]
LGTM

**Review by mposolda — APPROVED** [review]
Approving on behalf of @martin-kanis

**Pull request merged**